### PR TITLE
Update Step data saving infrastructure

### LINKED ISSUE #1307: cube_build not honoring output_dir
While running `calwebb_spec3` with `output_dir` specified, the `s3d` products appeared in the working directory.

### LINKED ISSUE #1322: Refactor and simplify file I/O for Step
In light of further development of the level 3 pipelines, inclusion of associations, and issues [such as this](https://github.com/STScI-JWST/jwst/pull/1316), a refactor of the Step file I/O is in order.

TMI, a "shower revelation" occurred: Up to this point, all work on the I/O has attempted to preserve all of the existing logic. However, much of the I/O API appears to have been designed before much of anything had been developed, except maybe `calwebb_detector1`. However, with everything being read and processed through `DataModel`s and associations, where only suffix changes occur, much of the current logic can be removed.  The remaining code can be refactor to take advantage of the `DataModel`/association input.

# Implementation 
- [ ] PR #1316 
- [x] Resolves #1016
- [x] Resolves #1288
- [x] Resolves #942 
- [x] Resolves #798 
- [x] #756 _previously resolved_

### PR DESCRIPTION
**Pull request by stscieisenhamer**
# Issue
Refactor a number of components in both `Step`s and `DataModel` related to file saving. Issues resolved are:

# Implementation
- [x] Resolves #1307 
- [x] Fix saving multiple results/`ModelContainer` results in the `Step` architecture. In particular, during `Spec3Pipeline` processing, properly write out multiple results for the `extract_1d` step when `cube_build` produces more than one result.
- [x] Resolves #1322 
- [x] `caldetector1` is not replacing known suffix `rate` when `output_file` is specified.
- [x] Remove `output_basename` and make everything use `output_file`
- [x] Add `fits` as a default extension since associations don't put one on.  Or maybe they should.
- [x] `Step.py::427`: remove `output_file` from the logic
- [x] Reconsider `Step._make_result_id`
- [x] template usage
    - [x]  `format_template`
        - add ability to handle lists
            _not necessary at the moment_
        - [x] add parameter to drop unused replacement strings.
    - [ ] update association generation
    - [x] update pipelines
        - [x] #1598 
        - [x] #1478
- [x] walk through all steps/pipelines for consistency
    - [x] `calwebb_tso3`
        - [x] `ModelContainer` and things simply appending
        - [x] `output_dir` not being honored. wtf
- [x] pandokia regression
    - [x] update local repo
    - [x] merge local repo back
    - [x] merge this PR
    - [x] ???

### CODE REVIEW

File: jwst/stpipe/step.py [hunk 1]
```diff
@@ -422,9 +425,16 @@ def run(self, *args):
                     'make_output_path', parent_first=True
                 )
                 for idx, result in enumerate(results_to_save):
```
Comment (jdavies-st) [inline jwst/stpipe/step.py]: Does this this handle the case were the step returns a `ModelContainer` where each `DataModel` in the container has its `meta.filename` attribute populated?  Does it use `meta.filename` or does it try to index.

This block is currently the bug that @jemorrison is seeing when she populates `meta.filename` of each model in the container and then returns the container at the end of the step.

Comment (stscieisenhamer) [reply jwst/stpipe/step.py]: For documentation purposes: This particular issue fixed in #1338

**Comment by stscieisenhamer** [comment]
hmm, not sure how this got closed...reopening...

**Comment by hbushouse** [comment]
@stscieisenhamer Things seem to be pretty stable right now on master, with nothing important coming up, so this might be a good time to try merging this (assuming you think it's ready to go).

**Pull request merged**